cilium/cilium
# hubble: Add Support for filtering on HTTP headers

Please ensure your pull request adheres to the following guidelines:

      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
      please add the commit author[s] as reviewer[s] to this issue.



Implements the server side changes for  https://github.com/cilium/hubble/issues/776

```release-note
hubble: Add Support for filtering on HTTP headers
```


### How to test this manually

1. Deploy cilium from source with `make kind-image && make kind-install-cilium` (note the `hubble-export-file-path` should be set for some weird reason that is not super clear to me, this is how the code flow works at https://github.com/cilium/cilium/blob/f078737ed9bc0ad38cee0f6df5dd9624d83f4aad/daemon/cmd/hubble.go#L173)
2. Enable hubble with `cilium hubble enable --chart-directory=/home/chrismark/go/src/github.com/cilium/cilium/install/kubernetes/cilium`
3. Apply a L7 example endpoint with `kubectl apply -f https://raw.githubusercontent.com/cilium/cilium/1.13.3/examples/minikube/sw_l3_l4_l7_policy.yaml`
4. Port forward hubble relay svc with `kubectl port-forward -n kube-system svc/hubble-relay --address 0.0.0.0 --address :: 4245:80`
5. Deploy the sample app with `kubectl create -f https://raw.githubusercontent.com/cilium/cilium/1.13.3/examples/minikube/http-sw-app.yaml`. From https://docs.cilium.io/en/stable/gettingstarted/demo/#starwars-demo

-> Build hubble CLI from https://github.com/cilium/hubble/pull/1277 after replacing the `cilium` dependency in the `go.mod` file with `replace github.com/cilium/cilium => /home/chrismark/go/src/github.com/cilium/cilium`.

7. Observe the flows with `./hubble observe --pod tiefighter --http-header Foo:bar -f`
8. Execute a *matching* curl request with `kubectl exec tiefighter -- curl -H "Foo: bar" -s -XPOST "http://deathstar.default.svc.cluster.local/v1/request-landing"`
9. Verify that the flow is observed:
```console
Oct 29 19:46:06.337: default/tiefighter:39348 (ID:49781) -> default/deathstar-7848d6c4d5-9z2wr:80 (ID:43688) http-request FORWARDED (HTTP/1.1 POST http://deathstar.default.svc.cluster.local/v1/request-landing)
```
10. Execute a *non-matching* curl request with `kubectl exec tiefighter -- curl -H "Foo: baz" -s -XPOST "http://deathstar.default.svc.cluster.local/v1/request-landing"`
11. Verify that the flow is not observed.
